mountpoint-s3/src/inode/readdir.rs

```diff
@@ -0,0 +1,454 @@
+//! Utilities for implementing the `readdir` operation on a directory inode.
+//!
+//! `readdir` is conceptually simple ("just call ListObjectsV2") but has some subtleties that make
+//! it complicated:
+//!
+//! 1. It's possible for a directory to contain both a subdirectory and a file of the same name, in
+//!    which case we need to implement shadowing in the right way.
+//! 2. A directory can also contain local files/subdirectories, which we want to have appear in the
+//!    readdir stream, but be shadowed by remote files/subdirectories.
+//! 3. ListObjectsV2 returns common prefixes including the trailing '/', which messes up ordering
+//!    of entries in the readdir stream. For example, `a-/` < `a/`, but `a-` > `a` in lexicographic
+//!    order, so we need to re-sort the common prefixes rather than just streaming them directly.
+//! 4. We want to do large ListObjectsV2 calls (i.e. with a large page size), but `readdir` calls
+//!    typically are much smaller, so we need to hold onto ListObjectsV2 results for a while. But we
```

Comment:
What do we mean by holding on to the results of ListObjectsV2 calls? Is it having more results than a readdir calls can return? And are we handling them using multiple iterations or something? It was not clear for me from the description of ReadDirHandleIterator. Maybe I need to understand the code a bit better.